include/bluetooth/mesh/scene_srv.h

```diff
@@ -0,0 +1,230 @@
+	 *
+	 *  The callback should fill the @c data with the current scene state
+	 *  and return the number of bytes written. @c data is guaranteed to
+	 *  fit @c maxlen number of bytes.
+	 *
+	 *  @param mod Model to get the scene data of.
+	 *  @param data Scene data buffer to fill. Fits @c maxlen bytes.
+	 *
+	 *  @return The number of bytes written to @c data or a negative value
+	 *          on failure.
+	 */
+	ssize_t (*store)(struct bt_mesh_model *mod, uint8_t data[]);
+
+	/** @brief Recall a scene based on the given scene data.
+	 *
+	 *  When a scene is recalled, the Scene Server calls this callback for
+	 *  every scene entry that has data for the recalled scene. The handler
+	 *  shall start transitioning to the given scene with the given
+	 *  transition parameters.
+	 *
+	 *  @param mod        Model to restore the scene of.
+	 *  @param data       Scene data to restore.
+	 *  @param len        Scene data length.
+	 *  @param transition Transition parameters.
+	 */
+	void (*recall)(struct bt_mesh_model *mod, const uint8_t data[],
+		       size_t len, struct bt_mesh_model_transition *transition);
+};
+
+/** @brief Scene entry.
+ *
+ *  Every model that stores data in scenes must own a unique scene entry, and
+ *  register it with a Scene Server through @ref bt_mesh_scene_entry_add.
+ *
+ *  Parameters in this structure will be filled by @ref bt_mesh_scene_entry_add,
+ *  and should not be manipulated directly.
+ */
+struct bt_mesh_scene_entry {
+#if defined(CONFIG_BT_MESH_SCENE_SRV)
+	/** Scene Server this entry got registered to. */
+	struct bt_mesh_scene_srv *srv;
+	/** Model this scene entry belongs to. */
+	struct bt_mesh_model *mod;
+	/** Scene entry callbacks */
+	const struct bt_mesh_scene_entry_type *type;
+	/** Scene entry list node */
+	sys_snode_t n;
+#endif
+};
+
+/** @brief Register a scene entry.
+ *
+ *  This function fills all fields of the supplied @ref bt_mesh_scene_entry
+ *  structure, and registers it to the correct Scene Server.
+ *
+ *  Scene Servers store scene data for all models in their own element, and
+ *  every subsequent element until the next Scene Server. If a Scene Server is
+ *  found for this entry, the @c srv parameter will point to it when this
+ *  function returns. If @c srv is NULL, this means no Scene Server was found,
+ *  and Scene data will not be stored for this entry.
+ *
+ *  @note This function must be called as part of the model initialization
+ *        procedure to correctly recover a scene on startup. The initial Scene
+ *        is recovered as part of the model start procedure.
+ *
+ *  @param[in] mod   Model this scene entry represents.
+ *  @param[in] entry Scene entry.
+ *  @param[in] type  Scene entry type.
+ *  @param[in] vnd   Whether this is a vendor model.
+ */
+void bt_mesh_scene_entry_add(struct bt_mesh_model *mod,
+			     struct bt_mesh_scene_entry *entry,
+			     const struct bt_mesh_scene_entry_type *type,
+			     bool vnd);
```

Comment:
I'm thinking... that the access layer actually knows whether the model is a vendor model or not. So since we have a model context here, instead of having `bool vnd` as a parameter in the API, we could ask the access layer if this model is a vendor model. However I see `bool vnd` is already used in some other API functions, so maybe we could continue following this style...

Reply:
The access layer doesn't, actually. It can find out by searching for the model in the element, but that would be a waste of CPU cycles for something that is known at compile time. This information could have been assigned to the model structure on init without wasting any space though, so I'm not really disagreeing with you. Like you point out, this is already the established way of conveying this, so I'd like to not diverge from the pattern.